data/beaconrestapi/src/main/java/tech/pegasys/teku/beaconrestapi/handlers/v1/events/EventSubscriber.java

```diff
@@ -0,0 +1,92 @@
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.beaconrestapi.handlers.v1.events;
+
+import io.javalin.http.sse.SseClient;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import tech.pegasys.teku.infrastructure.async.AsyncRunner;
+
+public class EventSubscriber {
+  private static final Logger LOG = LogManager.getLogger();
+  private static final int MAX_PENDING_EVENTS = 10;
+  private final List<EventType> eventTypes;
+  private final SseClient sseClient;
+  private final Queue<QueuedEvent> queuedEvents;
+  private final Runnable closeCallback;
+  private final AtomicBoolean processingQueue;
+  final AsyncRunner asyncRunner;
+
+  public EventSubscriber(
+      final List<String> eventTypes,
+      final SseClient sseClient,
+      final Runnable closeCallback,
+      final AsyncRunner asyncRunner) {
+    this.eventTypes = getTopics(eventTypes);
+    this.sseClient = sseClient;
+    this.closeCallback = closeCallback;
+    this.queuedEvents = new ConcurrentLinkedQueue<>();
+    this.processingQueue = new AtomicBoolean(false);
+    this.asyncRunner = asyncRunner;
+    this.sseClient.onClose(closeCallback);
+  }
+
+  public void onEvent(final EventType eventType, final String message) {
+    if (eventTypes.contains(eventType)) {
+      if (queuedEvents.size() < MAX_PENDING_EVENTS) {
+        queuedEvents.add(QueuedEvent.of(eventType, message));
+        processEventQueue();
+      } else {
+        LOG.trace("Closing client connection due to exceeding the pending message limit");
+        sseClient.ctx.req.getAsyncContext().complete();
+        closeCallback.run();
+      }
+    }
+  }
+
+  static List<EventType> getTopics(List<String> topics) {
+    return topics.stream().map(EventType::valueOf).collect(Collectors.toList());
+  }
+
+  public SseClient getSseClient() {
+    return sseClient;
+  }
+
+  private void processEventQueue() {
+    if (!processingQueue.compareAndSet(false, true)) {
+      // any queue processing in progress will clear the queue, no need to run another instance
+      return;
+    }
+    asyncRunner
+        .runAsync(
+            () -> {
+              LOG.trace(
+                  "Processing queue with {} elements for event client {}",
+                  queuedEvents.size(),
+                  sseClient.hashCode());
+              while (queuedEvents.size() > 0) {
+                QueuedEvent event = queuedEvents.peek();
+                queuedEvents.remove(event);
+                sseClient.sendEvent(event.getEventType().name(), event.getMessageData());
+              }
```

Comment:
Why not use `poll()`?
```suggestion
              while (queuedEvents.size() > 0) {
                QueuedEvent event = queuedEvents.poll();
                if (poll == null) {
                   break;
                }
                sseClient.sendEvent(event.getEventType().name(), event.getMessageData());
              }
```